### PR TITLE
created Simple Options Handler

### PR DESCRIPTION
**Pull request by t-keazirian**
* added tests for correct Allow header
* added private methods to add headers to allowed methods list

### CODE REVIEW

File: TKeazirian.HTTPServer/Server/Router.cs [hunk 1]
```diff
@@ -19,9 +19,17 @@ public Response Route(Request request)
...
-                : handler.HandleResponse(request);
+            if (request.GetRequestMethod() == "HEAD")
+            {
+                return HeadRequest(handler, request);
+            }
+
+            if (request.GetRequestMethod() == "OPTIONS" && request.GetRequestPath() != "/method_options2")
+            {
+                return OptionsRequest(handler, request);
+            }
```
Comment (Aoxian) [inline TKeazirian.HTTPServer/Server/Router.cs:30]: you have a hard-coded route here in the `!= "/method_options2"`.

File: TKeazirian.HTTPServer/Handler/SimpleOptionsHandler.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    private string AddHeadPutPostToAllowedMethods()
+    {
+        List<string> allowHead = new List<string>() { "HEAD", "OPTIONS", "PUT", "POST" };
+        allowHead.AddRange(AllowedHttpMethods());
+
+        string allowedMethods = string.Join(", ", allowHead);
+        return allowedMethods;
+    }
```
Comment (Aoxian) [inline TKeazirian.HTTPServer/Handler/SimpleOptionsHandler.cs:45]: I would not expect to see anything about HEAD or OPTIONS in this Handler at all.

File: TKeazirian.HTTPServer.Tests/Router/RouterTests.cs [hunk 3]
```diff
@@ -77,4 +77,93 @@ public void ResourceCannotBeFoundReturns404()
...
+    {
+        var testRoutesConfig = new RoutesConfig(new Dictionary<string, Handler>
+        {
+            { "/test_path", new MockHandler() },
+        });
+        Request testRequest = new Request("OPTIONS", "/test_path", "", "");
+
+        Router router = new Router(testRoutesConfig);
+
+        var newAllowedMethods = router.AddToAllowedMethodsForOptions(testRequest);
```
Comment (Aoxian) [inline TKeazirian.HTTPServer.Tests/Router/RouterTests.cs:147]: what is new about this variable?

**Review by Aoxian — COMMENTED** [review]
Looks like you are off to a good start with the tests you added.  I wonder about all the code specific to HEAD and OPTIONS in the Handlers though.

File: TKeazirian.HTTPServer/Handler/SimpleOptionsHandler.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        List<string> allowHead = new List<string>() { "HEAD, OPTIONS" };
+        allowHead.AddRange(AllowedHttpMethods());
+
+        string allowedMethods = string.Join(", ", allowHead);
+        return allowedMethods;
+    }
+
+    private string AddHeadPutPostToAllowedMethods()
+    {
+        List<string> allowHead = new List<string>() { "HEAD", "OPTIONS", "PUT", "POST" };
```
Comment (Aoxian) [inline TKeazirian.HTTPServer/Handler/SimpleOptionsHandler.cs:40]: If you are supporting PUT and POST, I would expect to see something building those responses in this Handler.

Comment (t-keazirian) [reply TKeazirian.HTTPServer/Handler/SimpleOptionsHandler.cs:40]: That's the confusing part.  With the acceptance tests, they only test that they are allowed but not that they do anything with them.  Another layer!

Comment (Aoxian) [reply TKeazirian.HTTPServer/Handler/SimpleOptionsHandler.cs:40]: time to build the some of the structure? let the tests guide you.

File: TKeazirian.HTTPServer/Server/Router.cs (hunk 1 above)
Comment (t-keazirian) [reply TKeazirian.HTTPServer/Server/Router.cs]: removed!

File: TKeazirian.HTTPServer/Handler/SimpleOptionsHandler.cs (hunk 4 above)
Comment (t-keazirian) [reply TKeazirian.HTTPServer/Handler/SimpleOptionsHandler.cs]: I have updated it - not sure if it's in the right direction but maybe?!

File: TKeazirian.HTTPServer.Tests/Router/RouterTests.cs [hunk 5]
```diff
@@ -77,4 +78,113 @@ public void ResourceCannotBeFoundReturns404()
...
+        Assert.Equal("HTTP/1.1 200 OK\r\n", response.ResponseStatusLine);
+        Assert.Equal(testHeaders, response.ResponseHeaders);
+        Assert.Empty(response.GetBody());
+    }
+
+    [Fact]
+    public void OptionsHandlerHasAllowHeaderWithHeadGetOptionsPutPostMethods()
+    {
+        Request testRequest = new Request("OPTIONS", "/method_options2", "", "");
+        // string testHeaders = "Allow: HEAD, OPTIONS, GET, PUT, POST\r\n\r\n";
```
Comment (Aoxian) [inline TKeazirian.HTTPServer.Tests/Router/RouterTests.cs:105]: 🔫 pew pew pew

File: TKeazirian.HTTPServer.Tests/Router/RouterTests.cs [hunk 6]
```diff
@@ -77,4 +78,113 @@ public void ResourceCannotBeFoundReturns404()
...
+
+        Assert.Equal("HTTP/1.1 200 OK\r\n", response.ResponseStatusLine);
+        Assert.Equal(testHeaders, response.ResponseHeaders);
+        Assert.Empty(response.GetBody());
+    }
+
+    [Fact]
+    public void OptionsHandlerHasAllowHeaderWithHeadGetOptionsPutPostMethods()
+    {
+        Request testRequest = new Request("OPTIONS", "/method_options2", "", "");
```
Comment (Aoxian) [inline TKeazirian.HTTPServer.Tests/Router/RouterTests.cs:104]: in all the other tests prior to this PR, you are using mocks.  Why introduce actual request calls and couple your tests to an implementation of configuration?  Shouldn't you expand the existing mock instead?

File: TKeazirian.HTTPServer.Tests/helpers/HelperFunctions.cs [hunk 7]
```diff
@@ -12,11 +12,11 @@ public static string StringTestRequest(string verb, string path, string body = "
         return testRequest;
     }
 
-    public static string CreateTestResponseHeaders()
+    public static string CreateTestResponseHeaders(string body)
     {
         string testResponseHeaders =
             $"Content-Type: text/plain{Constants.NewLine}" +
-            $"Content-Length: 11{Constants.NewLine}{Constants.NewLine}";
+            $"Content-Length: {body.Length}{Constants.NewLine}{Constants.NewLine}";
```
Comment (Aoxian) [inline TKeazirian.HTTPServer.Tests/helpers/HelperFunctions.cs:19]: nice

File: TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+    public override Response HandleResponse(Request request)
+    {
+        return new ResponseBuilder()
+            .SetStatusCode(HttpStatusCode.Ok)
+            .SetHeaders("Content-Type", "text/plain")
+            .SetHeaders("Content-Length", request.GetRequestBody().Length.ToString())
+            .SetBody(request.GetRequestBody())
+            .Build();
+    }
```
Comment (Aoxian) [inline TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs:21]: are all three methods just going to use the same handle response? this may work for now.  What happens when an endpoint needs to handle multiple methods though?

File: TKeazirian.HTTPServer/Server/Router.cs [hunk 9]
```diff
@@ -19,9 +19,17 @@ public Response Route(Request request)
...
+            {
+                return HeadRequest(handler, request);
+            }
+
+            if (request.GetRequestMethod() == "OPTIONS")
+            {
+                return OptionsRequest(request);
+            }
+
+            return handler.HandleResponse(request);
```
Comment (Aoxian) [inline TKeazirian.HTTPServer/Server/Router.cs:32]: this is nice and easy to understand.  Makes sense to me.

File: TKeazirian.HTTPServer/Server/Router.cs [hunk 10]
```diff
@@ -56,4 +64,31 @@ private Response HeadRequest(Handler handler, Request request)
...
+        return allowedMethodsString;
+    }
+
+    public string AddToAllowedMethodsForOptions(Request request)
+    {
+        string allowedMethods = GetAllowedMethodsFromHandler(_routesConfig.Routes[request.GetRequestPath()]);
+
+        string additionalAllowedMethods = "HEAD, OPTIONS";
+        return $"{allowedMethods}, {additionalAllowedMethods}";
+    }
```
Comment (Aoxian) [inline TKeazirian.HTTPServer/Server/Router.cs:93]: what happens if GET is not implemented on an endpoint?  should you still "get" HEAD & OPTIONS?  If not, should there be something enforcing that at a minimum all endpoints have a GET?

Comment (Aoxian) [reply TKeazirian.HTTPServer/Server/Router.cs:93]: ... even if that enforcing is just an update to the README?

File: TKeazirian.HTTPServer.Tests/Router/RouterTests.cs (hunk 6 above)
Comment (t-keazirian) [reply TKeazirian.HTTPServer.Tests/Router/RouterTests.cs:104]: So would I create another Mock that looks like the actual implementation?  I think sometimes I don't do Mocks because I worry then I won't be testing the actual thing and then I won't know for sure if it's working (if that makes sense...)

File: TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs (hunk 8 above)
Comment (t-keazirian) [reply TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs:21]: Ah I see...good point.

Comment (t-keazirian) [reply TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs:21]: In terms of `PUT` (from RFC: Replace all current representations of the target resource with the request payload.), I'm not sure how to approach it...I think `GET` can just return 200 but I don't think I am grasping how I would differentiate `PUT` from `POST`...any thoughts/hints?

File: TKeazirian.HTTPServer/Server/Router.cs (hunk 10 above)
Comment (t-keazirian) [reply TKeazirian.HTTPServer/Server/Router.cs:93]: Well, if GET is not implemented, can you have HEAD or OPTIONS?  Since you get them "for free" with GET, if they don't have GET, then do you still get them?

File: TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs (hunk 8 above)
Comment (Aoxian) [reply TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs]: It's probably not best to implement this in any way; what does PUT do when the request is properly received but fails? Why no just do that?

Comment (Aoxian) [reply TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs]: ref: [HTTP PUT request method](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT)

Comment (Aoxian) [reply TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs]: How about a [501](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/501) ?

File: TKeazirian.HTTPServer/Server/Router.cs (hunk 10 above)
Comment (Aoxian) [reply TKeazirian.HTTPServer/Server/Router.cs:93]: I think you have your answer.

File: TKeazirian.HTTPServer/Server/Router.cs [hunk 11]
```diff
@@ -19,9 +19,17 @@ public Response Route(Request request)
             IsHttpMethodAllowed(request))
         {
             var handler = _routesConfig.Routes[request.GetRequestPath()];
-            return request.GetRequestMethod() == "HEAD"
-                ? HeadRequest(handler, request)
-                : handler.HandleResponse(request);
+            if (request.GetRequestMethod() == "HEAD" && handler.AllowedHttpMethods().Contains("GET"))
```
Comment (t-keazirian) [inline TKeazirian.HTTPServer/Server/Router.cs:22]: Re: https://github.com/t-keazirian/HTTPServer/pull/27#discussion_r911368942

Would this be considered enforcing that it has a GET request if you are asking for an OPTIONS or HEAD request?

File: TKeazirian.HTTPServer.Tests/Router/RouterTests.cs [hunk 12]
```diff
@@ -166,41 +166,23 @@ public void IfOptionsMethodIsAllowedOnEndpointThenGivesAppropriateHeaders()
...
+    [InlineData("PUT")]
+    public void PostForSimpleOptionsReturns501NotImplemented(string method)
     {
-        Request testRequest = new Request("PUT", "/method_options2", "", "Mock body");
+        Request testRequest = new Request(method, "/method_options2", "", "Mock body");
         var testRoutesConfig = new RoutesConfig(new Dictionary<string, Handler>
         {
-            { "/test_path", new MockHandler() },
             { "/method_options2", new MockPostHandler() }
         });
```
Comment (Aoxian) [inline TKeazirian.HTTPServer.Tests/Router/RouterTests.cs:176]: Should this be updated to some generic endpoint? to avoid confusing with you actual implementation.  Looks like '/test_endpoint' is what you are using above.

File: TKeazirian.HTTPServer.Tests/Router/RouterTests.cs [hunk 13]
```diff
@@ -209,7 +191,6 @@ public void GetForSimpleOptionsReturns200Ok()
         Request testRequest = new Request("GET", "/method_options2", "", "");
         var testRoutesConfig = new RoutesConfig(new Dictionary<string, Handler>
         {
-            { "/test_path", new MockHandler() },
             { "/method_options2", new MockPostHandler() }
```
Comment (Aoxian) [inline TKeazirian.HTTPServer.Tests/Router/RouterTests.cs:192]: why did you change from the generic '/test_path' here?  perhaps name it '/mock_post_path' ?

Comment (Aoxian) [inline TKeazirian.HTTPServer.Tests/Router/RouterTests.cs:192]: why did you change from the generic '/test_path' here?  perhaps name it '/mock_post_path' ?

File: TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        return new List<string>() { "GET", "PUT", "POST" };
+    }
+
+    public override Response HandleResponse(Request request)
+    {
+        if (request.GetRequestMethod() == "PUT" || request.GetRequestMethod() == "POST")
+        {
+            return new ResponseBuilder()
+                .SetStatusCode(HttpStatusCode.NotImplemented)
```
Comment (tpetersen0308) [inline TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs:18]: What would happen if there was a PATCH or DELETE request to this handler? Do you think you could generalize the `NotImplemented` behavior for all unimplemented request methods?

Comment (t-keazirian) [reply TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs:18]: Well, what happens now is that it won't even enter that Handler if the request method isn't included in the AllowedMethods for that Handler.  So, if it is included (like PUT/POST here), you'll get a 501 and if it isn't included (like a PATCH or DELETE), it will give you a 404.  Are you saying you would prefer that behavior to change, in general?

Comment (t-keazirian) [reply TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs:18]: As in, should I be adding a handler for 501 Not Implemented and have that be the Handler called when the method isn't included in the AllowedHttpMethods, and then the 404 Not Found method is called in....other cases?  When the path doesn't exist, perhaps?

If so, would that be out of scope of this PR and added as a separate ticket?

File: TKeazirian.HTTPServer.Tests/Router/RouterTests.cs (hunk 13 above)
Comment (t-keazirian) [reply TKeazirian.HTTPServer.Tests/Router/RouterTests.cs]: Fixed!

File: TKeazirian.HTTPServer.Tests/Router/RouterTests.cs (hunk 12 above)
Comment (t-keazirian) [reply TKeazirian.HTTPServer.Tests/Router/RouterTests.cs]: Fixed!

File: TKeazirian.HTTPServer.Tests/Router/RouterTests.cs [hunk 15]
```diff
@@ -167,7 +165,9 @@ public void IfOptionsMethodIsAllowedOnEndpointThenGivesAppropriateHeaders()
     [Theory]
     [InlineData("POST")]
     [InlineData("PUT")]
-    public void PostForSimpleOptionsReturns501NotImplemented(string method)
+    [InlineData("DELETE")]
+    [InlineData("PATCH")]
+    public void SimpleOptionsReturns501NotImplementedWhenNotImplementedMethod(string method)
```
Comment (Aoxian) [inline TKeazirian.HTTPServer.Tests/Router/RouterTests.cs:170]: This is only for SimpleOptions? or for any given endpoint?

File: TKeazirian.HTTPServer/Response/HeadResponse.cs [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+        var getResponse = handler.HandleResponse(getRequest);
+
+        var headResponse = new Response(
+            getResponse.GetStatusLine(),
+            getResponse.GetHeaders(),
+            ""
+        );
+        return headResponse;
+    }
+}
```
Comment (Aoxian) [inline TKeazirian.HTTPServer/Response/HeadResponse.cs:25]: This is an observation; typically in statically types languages, the use of `get` and `set` in names has a different meaning than how you are using it here.  In other words `getRequest` connotes "getting a request" and not an http GET verb.  For variables names, try to use nouns and reserve the use of verbs for the function names (like `request.GetRequestPath()`

File: TKeazirian.HTTPServer/Router/Router.cs [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+            if (IsOptionsResponse(request) && HandlerContainsGetMethod(handler))
+            {
+                return new OptionsResponse().BuildOptionsResponse(request, _routesConfig);
+            }
+
+            return handler.HandleResponse(request);
+        }
+
+        return new ResourceNotFoundHandler().HandleResponse(request);
+    }
```
Comment (Aoxian) [inline TKeazirian.HTTPServer/Router/Router.cs:41]: I like this pattern.  You would want to guard first then work, so I suggest changing the order here.
```csharp
    public Response Route(Request request)
    {
        if !RouteConfigContainsPath(request)
        {
            return new ResourceNotFoundHandler().HandleResponse(request);
        }

        if !IsHttpMethodAllowed(request)
        {
            return new NotImplementedResponse().BuildNotImplementedResponse();
        }

        var handler = GetHandler(request);

        if (IsHeadRequest(request) && HandlerContainsGetMethod(handler))
        {
            return new HeadResponse().BuildHeadResponse(handler, request);
        }

        if (IsOptionsResponse(request) && HandlerContainsGetMethod(handler))
        {
            return new OptionsResponse().BuildOptionsResponse(request, _routesConfig);
        }

        return handler.HandleResponse(request);
    }
```

File: TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs (hunk 14 above)
Comment (tpetersen0308) [reply TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs]: To me this looks like it's introducing a bug into the code, where the client could get an incorrect/misleading error status (404, should be 501). So I think you need to handle it here, at least in some way.

How is 501 similar in nature to 404? How are you handling 404's generically in a way that you're not currently handling 501's? Could you handle 501's more like that?

Remember when we talked about how eventually, not having your HTTP methods be config would start to cause difficulties? I think we might be at that point now.

File: TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs [hunk 18]
```diff
@@ -0,0 +1,31 @@
+namespace TKeazirian.HTTPServer.Handler;
+
+using Response;
+using Request;
+
+public class SimpleOptionsHandler2 : Handler
```
[the remark this answers is not in the document]
Comment (tpetersen0308) [reply TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs:6]: I think this is another sign that not having your HTTP methods in route configuration is starting to cause some issues

File: TKeazirian.HTTPServer/Router/Router.cs [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+
+    private static bool IsOptionsResponse(Request request)
+    {
+        return request.GetRequestMethod() == "OPTIONS";
+    }
+
+    private static bool IsHeadRequest(Request request)
+    {
+        return request.GetRequestMethod() == "HEAD";
+    }
```
Comment (Aoxian) [inline TKeazirian.HTTPServer/Router/Router.cs:66]: when the name of the method and the one line command that is contained within are basically the same; ask yourself if the method is even required in the first place.

Is this just a lot of uneeded abstraction?

File: TKeazirian.HTTPServer/Response/HeadResponse.cs (hunk 16 above)
Comment (Aoxian) [reply TKeazirian.HTTPServer/Response/HeadResponse.cs:25]: its fine, I see what is happening, its just ever so slightly confusing. Would not let it be a blocker.

File: TKeazirian.HTTPServer/Router/Router.cs (hunk 19 above)
Comment (Aoxian) [reply TKeazirian.HTTPServer/Router/Router.cs:66]: a good example is:
`HandlerContainsGetMethod(Handler handler)`
`handler.AllowedHttpMethods().Contains("GET")`

They look almost identical, just slight changes in the order of the words.  In my opinion, this shows that you chose really good names for the methods and how they function.  Why not use them as is?

File: TKeazirian.HTTPServer/Router/Router.cs (hunk 17 above)
Comment (t-keazirian) [reply TKeazirian.HTTPServer/Router/Router.cs]: Okay I think I have implemented this change correctly!

File: TKeazirian.HTTPServer.Tests/Router/RouterTests.cs [hunk 20]
```diff
@@ -65,7 +65,8 @@ public void IfGetMethodIsAllowedOnEndpointThenHeadMethodIsAllowed()
     [Fact]
     public void ResourceCannotBeFoundReturns404()
     {
-        Request badTestRequest = new Request("", "GET", HelperFunctions.CreateTestResponseHeaders(), "/test_path");
+        Request badTestRequest =
+            new Request("", "GET", HelperFunctions.CreateTestResponseHeaders("blah"), "/test_path");
```
Comment (Aoxian) [inline TKeazirian.HTTPServer.Tests/Router/RouterTests.cs:69]: "blah" 😂

File: TKeazirian.HTTPServer/Router/Router.cs (hunk 19 above)
Comment (t-keazirian) [reply TKeazirian.HTTPServer/Router/Router.cs]: Got it - makes sense!!

File: TKeazirian.HTTPServer/Router/Router.cs [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+                return new HeadResponse().BuildHeadResponse(handler, request);
+            }
+
+            if (IsOptionsResponse(request, handler))
+            {
+                return new OptionsResponse().BuildOptionsResponse(request, _routesConfig);
+            }
+        }
+
+        return handler.HandleResponse(request);
```
Comment (t-keazirian) [inline TKeazirian.HTTPServer/Router/Router.cs:38]: @Aoxian I need it here too ([see](https://github.com/t-keazirian/HTTPServer/pull/27/files/afdccc7a4c47817bf115dc624d69f734f543575f#r915089579))

File: TKeazirian.HTTPServer/Router/Router.cs [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+
+    private static bool IsHeadRequest(Request request, Handler handler)
+    {
+        if (HandlerContainsGetMethod(handler))
+        {
+            return request.GetRequestMethod() == "HEAD";
+        }
+
+        return false;
+    }
```
Comment (Aoxian) [inline TKeazirian.HTTPServer/Router/Router.cs:79]: This method takes both a request and a handler; However you looking up the handler based on a request.
ex: `var handler = GetHandler(request)`.  Why not just pass in just the request and be done with it?
```suggestion
    private static bool IsHeadRequest(Request request)
    {
        if (HandlerContainsGetMethod(GetHandler(request))
        {
            return request.GetRequestMethod() == "HEAD";
        }

        return false;
    }
```

File: TKeazirian.HTTPServer/Router/Router.cs [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+    private readonly RoutesConfig _routesConfig;
+
+    public Router(RoutesConfig routesConfig)
+    {
+        _routesConfig = routesConfig;
+    }
+
+    public Response Route(Request request)
+    {
+        var handler = GetHandler(request);
```
[the remark this answers is not in the document]
Comment (Aoxian) [reply TKeazirian.HTTPServer/Router/Router.cs:18]: Nevermind, I see it used bellow to actually handle the request.

File: TKeazirian.HTTPServer/Router/Router.cs (hunk 22 above)
Comment (Aoxian) [reply TKeazirian.HTTPServer/Router/Router.cs:79]: I see why this would not be a good suggestion.  Looks good as is.

File: TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs (hunk 14 above)
Comment (t-keazirian) [reply TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs]: Yes,  I can definitely see that!  Can I punt it to complete this as is and then work on separate refactoring ticket to implement that?

Comment (t-keazirian) [reply TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs]: I also refactored it a bit more so I think it's clearer...

501 --> if the RoutesConfig contains the path, but the Method isn't Allowed (resource exists but that method isn't implemented)
404 --> if the RoutesConfig does NOT contain the path at all (resource doesn't exist)

But I agree that will definitely need a further refactor to introduce the Method/HTTP Verb along with the path and Handler...

File: TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+    }
+
+    public override Response HandleResponse(Request request)
+    {
+        if (request.GetRequestMethod() == "GET")
+        {
+            return new ResponseBuilder()
+                .SetStatusCode(HttpStatusCode.Ok)
+                .Build();
+        }
```
Comment (tpetersen0308) [inline TKeazirian.HTTPServer/Handler/SimpleOptionsHandler2.cs:20]: This is definitely more the kind of pattern I was looking for 👍

**Comment by tpetersen0308** [comment]
> > I could add in GET to all of the methods that you could potentially GET from but I'm not sure if that is the right approach?

I'd like to hear why YOU think this may or may not be a good approach

File: TKeazirian.HTTPServer/Server/Router.cs (hunk 10 above)
Comment (tpetersen0308) [reply TKeazirian.HTTPServer/Server/Router.cs]: Picking back up from my other comments below. What would be the purpose of adding GET to all endpoints? In theory, should you be able to have an endpoint that doesn't serve GET requests? What I've been trying to get at is that I would expect an OPTIONS request to return ALL of the implemented HTTP methods for ANY given endpoint in the server. So if an endpoint only has POST and PUT configured what would you expect the OPTIONS response to that endpoint to be? Should that be in scope for this ticket?

Comment (tpetersen0308) [reply TKeazirian.HTTPServer/Server/Router.cs]: So, what I'm asking is: why is GET privileged here? HEAD and OPTIONS are two different types of requests, so I think you need to answer two questions:

Does an endpoint need to serve GET requests to respond to a HEAD request?

Does an endpoint need to serve GET requests to respond to an OPTIONS request?

My understanding is that when you have a GET you get HEAD "for free," but you get OPTIONS "for free" for any endpoint that is configured with any methods, whether or not those methods include GET.

Comment (tpetersen0308) [reply TKeazirian.HTTPServer/Server/Router.cs]: https://stackoverflow.com/questions/60232424/using-http-head-method-when-the-underlying-ressource-does-not-allow-get

Comment (t-keazirian) [reply TKeazirian.HTTPServer/Server/Router.cs]: I see!  Okay that definitely makes sense.  Let me take a look at this from that perspective!

File: TKeazirian.HTTPServer/Router/Router.cs [hunk 25]
```diff
@@ -53,14 +58,9 @@ private bool RoutesConfigContainsPath(Request request)
         return _routesConfig.Routes.ContainsKey(request.GetRequestPath());
     }
 
-    private static bool IsOptionsRequest(Request request, Handler handler)
+    private static bool IsOptionsRequest(Request request)
     {
-        if (handler.AllowedHttpMethods().Contains("GET"))
```
Comment (t-keazirian) [inline TKeazirian.HTTPServer/Router/Router.cs]: @tpetersen0308 I removed this requirement here and it is now working as expected.

File: TKeazirian.HTTPServer/Router/Router.cs [hunk 26]
```diff
@@ -19,6 +19,11 @@ public Response Route(Request request)
 
         if (RoutesConfigContainsPath(request) && !IsHttpMethodAllowed(request))
         {
+            if (IsOptionsRequest(request))
+            {
+                return new OptionsResponse().BuildOptionsResponse(request, _routesConfig);
+            }
```
Comment (tpetersen0308) [inline TKeazirian.HTTPServer/Router/Router.cs:25]: This looks like it's the same code as lines 37-40. Do you need it in both places?

File: TKeazirian.HTTPServer/Router/Router.cs [hunk 27]
```diff
@@ -0,0 +1,87 @@
...
+        return false;
+    }
+
+    private bool IsHttpMethodAllowed(Request request)
+    {
+        Handler handler = _routesConfig.Routes[request.GetRequestPath()];
+
+        if (IsHeadRequest(request, handler) || IsOptionsRequest(request))
+        {
+            return handler.AllowedHttpMethods().Contains("GET");
```
[the remark this answers is not in the document]
Comment (t-keazirian) [reply TKeazirian.HTTPServer/Router/Router.cs:82]: Just saw that - need to remove the second half of the OR!

File: TKeazirian.HTTPServer/Router/Router.cs (hunk 26 above)
Comment (t-keazirian) [reply TKeazirian.HTTPServer/Router/Router.cs:25]: I did notice that duplication, as well!  Working on fix right now!

File: TKeazirian.HTTPServer/Router/Router.cs (hunk 27 above)
[the remark this answers is not in the document]
Comment (tpetersen0308) [reply TKeazirian.HTTPServer/Router/Router.cs:82]: Okay, cool! Then that seems like it should decouple OPTIONS from specific HTTP methods 👍
